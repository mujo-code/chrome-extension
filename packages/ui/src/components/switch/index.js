import { omitKeys } from '@mujo/box'
import { Functional } from '@mujo/utils'
import React from 'react'
import { Box } from '../box'
import { useTheme } from '../../hooks/use-theme'

const { noop } = Functional

const radius = {
  m: 16,
  l: 24,
}

export const Switch = props => {
  const { size = 'l', value, onChange = noop, borderSize = 1 } = props
  const otherProps = omitKeys(props, 'size', 'value', 'onChange')
  const { highlight, foreground, background } = useTheme()
  return (
    <Box
      {...otherProps}
      borderRadius={size}
      backgroundColor={highlight}
      display="flex"
      css={{ height: radius[size], width: radius[size] * 2 }}
      onClick={e => {
        e.target.value = !value
        onChange(e)
      }}
    >
      <Box
        borderRadius={size}
        backgroundColor={value ? foreground : background}
        margin
        css={{
          transition: 'all 0.3s',
          width: radius[size] - borderSize * 2,
          height: radius[size] - borderSize * 2,
          transform: `translate(${
            value ? radius[size] + borderSize : borderSize
          }px, 1px)`,
        }}
      />
    </Box>
  )
}
