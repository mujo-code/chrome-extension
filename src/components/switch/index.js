import { Box } from '@jcblw/box'
import { removeKeys } from '@jcblw/box/dist/lib/remove-keys'
import React from 'react'
import { useTheme } from '../../hooks/use-theme'
import { noop } from '../../lib/functional'

const radius = {
  m: 16,
  l: 24,
}

export const Switch = props => {
  const { size = 'l', value, onChange = noop, borderSize = 1 } = props
  const otherProps = removeKeys(props, 'size', 'value', 'onChange')
  const { highlight, foreground, background } = useTheme()
  return (
    <Box
      {...otherProps}
      borderRadius={size}
      backgroundColor={highlight}
      display="flex"
      css={{ height: radius[size], width: radius[size] * 2 }}
      onClick={e => {
        e.target.value = !value
        onChange(e)
      }}
    >
      <Box
        borderRadius={size}
        backgroundColor={value ? foreground : background}
        margin
        css={{
          transition: 'all 0.3s',
          width: radius[size] - borderSize * 2,
          height: radius[size] - borderSize * 2,
          transform: `translate(${
            value ? radius[size] + borderSize : borderSize
          }px, 1px)`,
        }}
      />
    </Box>
  )
}
