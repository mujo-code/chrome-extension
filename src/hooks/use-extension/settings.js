import {
  SUB_DETAILS_MODAL,
  CURRENT_SUB_SKU,
  SUPPORT_URL,
  BREATH_MIN,
  BREATH_MAX,
  WEBSTORE_URL,
} from '../../constants'
import { VERSION } from '../../env'

export const useSettings = ({
  user,
  setUpsellModal,
  alarmEnabled,
  setAlarmEnabled,
  hasPermission,
  requestPermissions,
  removePermissions,
  breathAmount,
  setBreathAmount,
  t,
}) => [
  {
    label: t('subscribe'),
    type: 'button',
    value: t('more-info'),
    alt: user.isSubscribed
      ? t('thanks-support')
      : t('get-more-access'),
    setter: () => {
      // always open this modal
      setUpsellModal({
        name: SUB_DETAILS_MODAL,
        sku: CURRENT_SUB_SKU,
        callback: () => {},
      })
    },
  },
  {
    label: t('reminder'),
    type: 'boolean',
    alt: t('reminder-alt'),
    setter: enabled => setAlarmEnabled(enabled),
    value: alarmEnabled,
  },
  {
    label: t('screen-time-enabled'),
    type: 'boolean',
    alt: t('screen-time-permissions'),
    setter: () => {
      if (hasPermission) {
        removePermissions()
      } else {
        requestPermissions()
      }
    },
    value: hasPermission,
  },
  {
    label: t('breath-amount'),
    type: 'number',
    alt: t('want-more-breath'),
    inputLabel: t('breath-limits', {
      min: BREATH_MIN,
      max: BREATH_MAX,
    }),
    setter: amount => {
      setBreathAmount(
        Math.min(Math.max(amount, BREATH_MIN), BREATH_MAX)
      )
    },
    value: breathAmount,
  },
  {
    label: t('help'),
    alt: t('join-chat'),
    type: 'button',
    value: t('get-support'),
    setter: () => {
      window.open(SUPPORT_URL)
    },
  },
  {
    label: t('about-mujo'),
    alt: t('mujo-version', { version: VERSION }),
    type: 'button',
    value: t('rate-us'),
    setter: () => {
      window.open(WEBSTORE_URL)
    },
  },
]
