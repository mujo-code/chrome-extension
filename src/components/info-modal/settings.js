import { Box } from '@mujo/box'
import React from 'react'
import { Button } from '../button'
import { HeaderS, BodyS } from '../fonts'
import { Input } from '../input'
import { Switch } from '../switch'

export const Boolean = ({
  onChange,
  value,
  background,
  color,
  highlight,
}) => {
  const change = e => onChange(e.target.value)
  return <Switch value={value} onChange={change} />
}

export const SettingsInput = props => {
  const { type, onChange, value, inputLabel } = props
  switch (type) {
    case 'button':
      return (
        <Button
          whiteSpace="nowrap"
          backgroundColor={props.highlight}
          color={props.foreground}
          onClick={onChange}
        >
          {value}
        </Button>
      )
    case 'number':
      return (
        <Input
          label={inputLabel || ' '}
          type="number"
          value={value}
          onChange={e =>
            onChange(Math.abs(parseInt(e.target.value, 10)))
          }
        />
      )
    case 'boolean':
    default:
      return <Boolean {...props} />
  }
}

export const SettingItem = ({
  type,
  alt,
  label,
  setter,
  value,
  theme,
}) => (
  <Box
    display="flex"
    direction="row"
    paddingTop="s"
    paddingBottom="s"
    borderBottom="s"
  >
    <Box
      backgroundColor={theme.backgroundSecondary}
      css={{ width: '4px' }}
      borderRadius="s"
      marginRight="m"
    />
    <Box flex="1" paddingRight="m" display="flex" direction="column">
      <HeaderS
        marginTop="zero"
        marginBottom={alt ? 'xs' : 'zero'}
        color={theme.foreground}
        flex="0"
      >
        {label}
      </HeaderS>
      {alt && (
        <BodyS
          marginTop="zero"
          marginBottom="zero"
          color={theme.foregroundSecondary}
        >
          {alt}
        </BodyS>
      )}
    </Box>
    <Box
      flex="0"
      display="flex"
      direction="column"
      alignItems={type === 'button' ? 'center' : 'flexStart'}
    >
      <SettingsInput
        type={type}
        value={value}
        onChange={setter}
        {...theme}
      />
    </Box>
  </Box>
)

export const settingsModal = (_, __, { settings, theme }) => ({
  title: 'Settings',
  children: (
    <>
      {settings.map((setting, i, arr) => (
        <SettingItem key={setting.label} {...setting} theme={theme} />
      ))}
    </>
  ),
})
