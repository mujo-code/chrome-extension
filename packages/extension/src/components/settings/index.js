import { Setting, useStorage } from '@mujo/plugins'
import React, { useCallback } from 'react'
import { useTranslation } from 'react-i18next'
import {
  SUB_DETAILS_MODAL,
  CURRENT_SUB_SKU,
  TRANSLATION_FILE,
  BREATH_MAX,
  BREATH_MIN,
  SUPPORT_URL,
  WEBSTORE_URL,
  VERSION,
  CONDITIONAL_NEW_TAB_PAGE_KEY,
} from '../../constants'
import { useExtension } from '../../hooks/use-extension'

export const Settings = ({ children }) => {
  const {
    user,
    setUpsellModal,
    setBreathAmount,
    breathAmount,
  } = useExtension()
  const { t } = useTranslation(TRANSLATION_FILE)
  const openUpsell = useCallback(() => {
    setUpsellModal({
      name: SUB_DETAILS_MODAL,
      sku: CURRENT_SUB_SKU,
      callback: () => {},
    })
  }, [setUpsellModal])

  const setBreath = useCallback(amount => {
    setBreathAmount(
      Math.min(Math.max(amount, BREATH_MIN), BREATH_MAX)
    )
  }, [])
  const openWebstore = useCallback(() => {
    window.open(WEBSTORE_URL)
  }, [])
  const openSupport = useCallback(() => {
    window.open(SUPPORT_URL)
  }, [])

  const [
    conditionalNewTabPageEnabled,
    setConditionalNewTabPageEnabled,
  ] = useStorage(CONDITIONAL_NEW_TAB_PAGE_KEY)
  return (
    <>
      <Setting
        label={t('subscribe')}
        type="button"
        value={t('more-info')}
        alt={
          user.isSubscribed
            ? t('thanks-support')
            : t('get-more-access')
        }
        onChange={openUpsell}
      />
      {children}
      <Setting
        label={t('conditional-new-tab-page')}
        type="boolean"
        alt={t('conditional-new-tab-page-alt')}
        onChange={setConditionalNewTabPageEnabled}
        value={conditionalNewTabPageEnabled}
      />
      <Setting
        label={t('breath-amount')}
        type="number"
        alt={t('want-more-breath', {
          min: BREATH_MIN,
          max: BREATH_MAX,
        })}
        onChange={setBreath}
        value={breathAmount}
      />
      <Setting
        label={t('help')}
        type="button"
        alt={t('join-chat')}
        onChange={openSupport}
        value={t('get-support')}
      />
      <Setting
        label={t('about-mujo')}
        type="button"
        alt={t('mujo-version', { version: VERSION })}
        onChange={openWebstore}
        value={t('rate-us')}
      />
    </>
  )
}
