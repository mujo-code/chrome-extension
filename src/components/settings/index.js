import { Setting } from '@mujo/plugins'
import React, { useCallback } from 'react'
import { useTranslation } from 'react-i18next'
import {
  SUB_DETAILS_MODAL,
  CURRENT_SUB_SKU,
  TRANSLATION_FILE,
  BREATH_MAX,
  BREATH_MIN,
  SUPPORT_URL,
  WEBSTORE_URL,
  VERSION,
} from '../../constants'
import { useExtension } from '../../hooks/use-extension'

export const Settings = ({ children }) => {
  const {
    user,
    setUpsellModal,
    setBreathAmount,
    breathAmount,
  } = useExtension()
  const { t } = useTranslation(TRANSLATION_FILE)
  const openUpsell = useCallback(() => {
    setUpsellModal({
      name: SUB_DETAILS_MODAL,
      sku: CURRENT_SUB_SKU,
      callback: () => {},
    })
  }, [setUpsellModal])

  const setBreath = useCallback(amount => {
    setBreathAmount(
      Math.min(Math.max(amount, BREATH_MIN), BREATH_MAX)
    )
  }, [])
  const openWebstore = useCallback(() => {
    window.open(WEBSTORE_URL)
  }, [])
  const openSupport = useCallback(() => {
    window.open(SUPPORT_URL)
  }, [])

  return (
    <>
      <Setting
        label={t('subscribe')}
        type="button"
        value={t('more-info')}
        alt={
          user.isSubscribed
            ? t('thanks-support')
            : t('get-more-access')
        }
        onChange={openUpsell}
      />
      {children}
      <Setting
        label={t('breath-amount')}
        type="number"
        alt={t('want-more-breath', {
          min: BREATH_MIN,
          max: BREATH_MAX,
        })}
        onChange={setBreath}
        value={breathAmount}
      />
      <Setting
        label={t('help')}
        type="button"
        alt={t('join-chat')}
        onChange={openSupport}
        value={t('get-support')}
      />
      <Setting
        label={t('about-mujo')}
        type="button"
        alt={t('mujo-version', { version: VERSION })}
        onChange={openWebstore}
        value={t('rate-us')}
      />
    </>
  )
}
