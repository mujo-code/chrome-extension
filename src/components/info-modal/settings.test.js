import React from 'react'
import { create } from 'react-test-renderer'
import { Switch } from '../switch'
import { settingsModal, SettingItem, Boolean } from './settings'

test('settingsModal should return an object', () => {
  const settings = [{ label: 'foo' }]
  expect(
    settingsModal(null, null, { settings, theme: {} }).title
  ).toBe('Settings')
})

test('SettingsItem should match snapshot', () => {
  const tree = create(
    <SettingItem label="foo" alt="bar" theme={{}} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SettingsItem boolean type should match snapshot', () => {
  const tree = create(
    <SettingItem
      type="boolean"
      label="foo"
      alt="bar"
      value="baz"
      theme={{}}
    />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('SettingsItem button type should match snapshot', () => {
  const tree = create(
    <SettingItem
      type="button"
      label="foo"
      alt="bar"
      value="baz"
      theme={{}}
    />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Boolean should unshell the value from switch on change', () => {
  const onChange = jest.fn()
  const eventMock = { target: { value: 'foo' } }
  const { root } = create(
    <Boolean value={true} onChange={onChange} />
  )
  const sw = root.findByType(Switch)
  sw.props.onChange(eventMock)
  expect(onChange).toBeCalledWith('foo')
})
