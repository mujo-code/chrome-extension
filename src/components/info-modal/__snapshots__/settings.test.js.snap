// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsItem boolean type should match snapshot 1`] = `
<div
  borderBottom="s"
  className="css-12jtym5"
>
  <div
    className="css-p1yvl5"
  />
  <div
    className="css-kjj7d4"
  >
    <h4
      className="css-ddfbvk"
      color="color"
    >
      foo
    </h4>
    <p
      className="css-xkhwk8"
      color="color"
    >
      bar
    </p>
  </div>
  <div
    className="css-5hlt78"
  >
    <div
      backgroundColor="saltBox"
      className="css-163c4dr"
      onClick={[Function]}
    >
      <div
        backgroundColor="mischka"
        className="css-4zx0ao"
      />
    </div>
  </div>
</div>
`;

exports[`SettingsItem button type should match snapshot 1`] = `
<div
  borderBottom="s"
  className="css-12jtym5"
>
  <div
    className="css-p1yvl5"
  />
  <div
    className="css-kjj7d4"
  >
    <h4
      className="css-ddfbvk"
      color="color"
    >
      foo
    </h4>
    <p
      className="css-xkhwk8"
      color="color"
    >
      bar
    </p>
  </div>
  <div
    className="css-180tdcr"
  >
    <button
      className="css-1akodnm"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      outlineColor="saltBox"
      position="relative"
    >
      baz
    </button>
  </div>
</div>
`;

exports[`SettingsItem should match snapshot 1`] = `
<div
  borderBottom="s"
  className="css-12jtym5"
>
  <div
    className="css-p1yvl5"
  />
  <div
    className="css-kjj7d4"
  >
    <h4
      className="css-ddfbvk"
      color="color"
    >
      foo
    </h4>
    <p
      className="css-xkhwk8"
      color="color"
    >
      bar
    </p>
  </div>
  <div
    className="css-5hlt78"
  >
    <div
      backgroundColor="saltBox"
      className="css-163c4dr"
      onClick={[Function]}
    >
      <div
        backgroundColor="outerSpace"
        className="css-dx4x1i"
      />
    </div>
  </div>
</div>
`;
